File: app/src/main/java/n7/ad2/news/db/NewsDao.java
package n7.ad2.news.db;

import androidx.lifecycle.LiveData;
import androidx.paging.DataSource;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import java.util.List;

@Dao
public interface NewsDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void setNews(NewsModel news);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void setNews(List<NewsModel> news);

    @Query("DELETE FROM NewsModel")
    void deleteAll();

    @Query("SELECT * FROM NewsModel")
    DataSource.Factory<Integer, NewsModel> getDataSourceNews();

    @Query("UPDATE NewsModel SET content=:s where href=:href")
    void setContent(String s, String href);

    @Query("SELECT * FROM NewsModel WHERE href =:href")
    LiveData<NewsModel> getNewsByIndex(String href);
}

File: core/database/src/main/kotlin/n7/ad2/database_guides/api/dao/NewsDao.kt
package n7.ad2.database_guides.api.dao

import androidx.paging.PagingSource
import androidx.room.Dao
import androidx.room.Query
import n7.ad2.database_guides.internal.model.NewsDB

@Dao
interface NewsDao : BaseDao<NewsDB> {

    @Query("SELECT * FROM NewsDB")
    fun getPagingSourceNews(): PagingSource<Int, NewsDB>

    @Query("SELECT * FROM NewsDB WHERE loadedFromPage =:fromPage")
    suspend fun getNews(fromPage: Int): List<NewsDB>

    @Query("DELETE FROM NewsDB")
    suspend fun clearAll()

    @Query("SELECT * FROM NewsDB WHERE rowId =:id")
    suspend fun getSingleNews(id: Int): NewsDB
}
